File: app/src/test/java/com/vrem/wifianalyzer/settings/SettingsTest.java
/*
 * WiFiAnalyzer
 * Copyright (C) 2019  VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */

package com.vrem.wifianalyzer.settings;

import android.content.SharedPreferences.OnSharedPreferenceChangeListener;

import com.vrem.util.EnumUtils;
import com.vrem.util.LocaleUtils;
import com.vrem.wifianalyzer.R;
import com.vrem.wifianalyzer.navigation.NavigationMenu;
import com.vrem.wifianalyzer.wifi.accesspoint.AccessPointViewType;
import com.vrem.wifianalyzer.wifi.accesspoint.ConnectionViewType;
import com.vrem.wifianalyzer.wifi.band.WiFiBand;
import com.vrem.wifianalyzer.wifi.graphutils.GraphLegend;
import com.vrem.wifianalyzer.wifi.model.GroupBy;
import com.vrem.wifianalyzer.wifi.model.Security;
import com.vrem.wifianalyzer.wifi.model.SortBy;
import com.vrem.wifianalyzer.wifi.model.Strength;

import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.Collections;
import java.util.HashSet;
import java.util.Locale;
import java.util.Set;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class SettingsTest {
    @Mock
    private Repository repository;
    @Mock
    private OnSharedPreferenceChangeListener onSharedPreferenceChangeListener;

    private Settings fixture;

    @Before
    public void setUp() {
        fixture = new Settings(repository);
    }

    @After
    public void tearDown() {
        verifyNoMoreInteractions(repository);
        verifyNoMoreInteractions(onSharedPreferenceChangeListener);
    }

    @Test
    public void testInitializeDefaultValues() {
        fixture.initializeDefaultValues();
        verify(repository).initializeDefaultValues();
    }

    @Test
    public void testRegisterOnSharedPreferenceChangeListener() {
        // execute
        fixture.registerOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener);
        // validate
        verify(repository).registerOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener);
    }

    @Test
    public void testGetScanSpeed() {
        // setup
        int defaultValue = 10;
        int expected = 3;
        when(repository.getStringAsInteger(R.string.scan_speed_default, Settings.SCAN_SPEED_DEFAULT)).thenReturn(defaultValue);
        when(repository.getStringAsInteger(R.string.scan_speed_key, defaultValue)).thenReturn(expected);
        // execute
        int actual = fixture.getScanSpeed();
        // validate
        assertEquals(expected, actual);
        verify(repository).getStringAsInteger(R.string.scan_speed_default, Settings.SCAN_SPEED_DEFAULT);
        verify(repository).getStringAsInteger(R.string.scan_speed_key, defaultValue);
    }

    @Test
    public void testIsWiFiThrottleDisabled() {
        // execute
        boolean actual = fixture.isWiFiThrottleDisabled();
        // validate
        assertFalse(actual);
    }


    @Test
    public void testGetGraphMaximumY() {
        // setup
        int defaultValue = 1;
        int value = 2;
        int expected = value * Settings.GRAPH_Y_MULTIPLIER;
        when(repository.getStringAsInteger(R.string.graph_maximum_y_default, Settings.GRAPH_Y_DEFAULT)).thenReturn(defaultValue);
        when(repository.getStringAsInteger(R.string.graph_maximum_y_key, defaultValue)).thenReturn(value);
        // execute
        int actual = fixture.getGraphMaximumY();
        // validate
        assertEquals(expected, actual);
        verify(repository).getStringAsInteger(R.string.graph_maximum_y_default, Settings.GRAPH_Y_DEFAULT);
        verify(repository).getStringAsInteger(R.string.graph_maximum_y_key, defaultValue);
    }

    @Test
    public void testGetGroupBy() {
        // setup
        when(repository.getStringAsInteger(R.string.group_by_key, GroupBy.NONE.ordinal())).thenReturn(GroupBy.CHANNEL.ordinal());
        // execute
        GroupBy actual = fixture.getGroupBy();
        // validate
        assertEquals(GroupBy.CHANNEL, actual);
        verify(repository).getStringAsInteger(R.string.group_by_key, GroupBy.NONE.ordinal());
    }

    @Test
    public void testGetSortBy() {
        // setup
        when(repository.getStringAsInteger(R.string.sort_by_key, SortBy.STRENGTH.ordinal())).thenReturn(SortBy.SSID.ordinal());
        // execute
        SortBy actual = fixture.getSortBy();
        // validate
        assertEquals(SortBy.SSID, actual);
        verify(repository).getStringAsInteger(R.string.sort_by_key, SortBy.STRENGTH.ordinal());
    }

    @Test
    public void testGetAccessPointView() {
        // setup
        when(repository.getStringAsInteger(R.string.ap_view_key, AccessPointViewType.COMPLETE.ordinal())).thenReturn(AccessPointViewType.COMPACT.ordinal());
        // execute
        AccessPointViewType actual = fixture.getAccessPointView();
        // validate
        assertEquals(AccessPointViewType.COMPACT, actual);
        verify(repository).getStringAsInteger(R.string.ap_view_key, AccessPointViewType.COMPLETE.ordinal());
    }

    @Test
    public void testGetConnectionViewType() {
        // setup
        when(repository.getStringAsInteger(R.string.connection_view_key, ConnectionViewType.COMPLETE.ordinal())).thenReturn(ConnectionViewType.COMPACT.ordinal());
        // execute
        ConnectionViewType actual = fixture.getConnectionViewType();
        // validate
        assertEquals(ConnectionViewType.COMPACT, actual);
        verify(repository).getStringAsInteger(R.string.connection_view_key, ConnectionViewType.COMPLETE.ordinal());
    }

    @Test
    public void testGetThemeStyle() {
        // setup
        when(repository.getStringAsInteger(R.string.theme_key, ThemeStyle.DARK.ordinal())).thenReturn(ThemeStyle.LIGHT.ordinal());
        // execute
        ThemeStyle actual = fixture.getThemeStyle();
        // validate
        assertEquals(ThemeStyle.LIGHT, actual);
        verify(repository).getStringAsInteger(R.string.theme_key, ThemeStyle.DARK.ordinal());
    }

    @Test
    public void testGetChannelGraphLegend() {
        // setup
        when(repository.getStringAsInteger(R.string.channel_graph_legend_key, GraphLegend.HIDE.ordinal())).thenReturn(GraphLegend.RIGHT.ordinal());
        // execute
        GraphLegend actual = fixture.getChannelGraphLegend();
        // validate
        assertEquals(GraphLegend.RIGHT, actual);
        verify(repository).getStringAsInteger(R.string.channel_graph_legend_key, GraphLegend.HIDE.ordinal());
    }

    @Test
    public void testGetTimeGraphLegend() {
        // setup
        when(repository.getStringAsInteger(R.string.time_graph_legend_key, GraphLegend.LEFT.ordinal())).thenReturn(GraphLegend.RIGHT.ordinal());
        // execute
        GraphLegend actual = fixture.getTimeGraphLegend();
        // validate
        assertEquals(GraphLegend.RIGHT, actual);
        verify(repository).getStringAsInteger(R.string.time_graph_legend_key, GraphLegend.LEFT.ordinal());
    }

    @Test
    public void testGetWiFiBand() {
        // setup
        when(repository.getStringAsInteger(R.string.wifi_band_key, WiFiBand.GHZ2.ordinal())).thenReturn(WiFiBand.GHZ5.ordinal());
        // execute
        WiFiBand actual = fixture.getWiFiBand();
        // validate
        assertEquals(WiFiBand.GHZ5, actual);
        verify(repository).getStringAsInteger(R.string.wifi_band_key, WiFiBand.GHZ2.ordinal());
    }

    @Test
    public void testGetSSIDFilter() {
        // setup
        Set<String> expected = new HashSet<>(Arrays.asList("value1", "value2", "value3"));
        when(repository.getStringSet(R.string.filter_ssid_key, Collections.emptySet())).thenReturn(expected);
        // execute
        Set<String> actual = fixture.getSSIDs();
        // validate
        assertEquals(expected, actual);
        verify(repository).getStringSet(R.string.filter_ssid_key, Collections.emptySet());
    }

    @Test
    public void testSaveSSIDFilter() {
        // setup
        Set<String> values = new HashSet<>(Arrays.asList("value1", "value2", "value3"));
        // execute
        fixture.saveSSIDs(values);
        // validate
        verify(repository).saveStringSet(R.string.filter_ssid_key, values);
    }

    @Test
    public void testGetWiFiBandFilter() {
        // setup
        WiFiBand expected = WiFiBand.GHZ5;
        Set<String> values = Collections.singleton("" + expected.ordinal());
        Set<String> defaultValues = EnumUtils.ordinals(WiFiBand.class);
        when(repository.getStringSet(R.string.filter_wifi_band_key, defaultValues)).thenReturn(values);
        // execute
        Set<WiFiBand> actual = fixture.getWiFiBands();
        // validate
        assertEquals(1, actual.size());
        assertTrue(actual.contains(expected));
        verify(repository).getStringSet(R.string.filter_wifi_band_key, defaultValues);
    }

    @Test
    public void testSaveWiFiBandFilter() {
        // setup
        Set<WiFiBand> values = Collections.singleton(WiFiBand.GHZ5);
        Set<String> expected = Collections.singleton("" + WiFiBand.GHZ5.ordinal());
        // execute
        fixture.saveWiFiBands(values);
        // validate
        verify(repository).saveStringSet(R.string.filter_wifi_band_key, expected);
    }

    @Test
    public void testGetStrengthFilter() {
        // setup
        Strength expected = Strength.THREE;
        Set<String> values = Collections.singleton("" + expected.ordinal());
        Set<String> defaultValues = EnumUtils.ordinals(Strength.class);
        when(repository.getStringSet(R.string.filter_strength_key, defaultValues)).thenReturn(values);
        // execute
        Set<Strength> actual = fixture.getStrengths();
        // validate
        assertEquals(1, actual.size());
        assertTrue(actual.contains(expected));
        verify(repository).getStringSet(R.string.filter_strength_key, defaultValues);
    }

    @Test
    public void testSaveStrengthFilter() {
        // setup
        Set<Strength> values = Collections.singleton(Strength.TWO);
        Set<String> expected = Collections.singleton("" + Strength.TWO.ordinal());
        // execute
        fixture.saveStrengths(values);
        // validate
        verify(repository).saveStringSet(R.string.filter_strength_key, expected);
    }

    @Test
    public void testGetSecurityFilter() {
        // setup
        Security expected = Security.WPA;
        Set<String> values = Collections.singleton("" + expected.ordinal());
        Set<String> defaultValues = EnumUtils.ordinals(Security.class);
        when(repository.getStringSet(R.string.filter_security_key, defaultValues)).thenReturn(values);
        // execute
        Set<Security> actual = fixture.getSecurities();
        // validate
        assertEquals(1, actual.size());
        assertTrue(actual.contains(expected));
        verify(repository).getStringSet(R.string.filter_security_key, defaultValues);
    }

    @Test
    public void testSaveSecurityFilter() {
        // setup
        Set<Security> values = Collections.singleton(Security.WEP);
        Set<String> expected = Collections.singleton("" + Security.WEP.ordinal());
        // execute
        fixture.saveSecurities(values);
        // validate
        verify(repository).saveStringSet(R.string.filter_security_key, expected);
    }

    @Test
    public void testToggleWiFiBand() {
        // setup
        when(repository.getStringAsInteger(R.string.wifi_band_key, WiFiBand.GHZ2.ordinal())).thenReturn(WiFiBand.GHZ5.ordinal());
        // execute
        fixture.toggleWiFiBand();
        // validate
        verify(repository).getStringAsInteger(R.string.wifi_band_key, WiFiBand.GHZ2.ordinal());
        verify(repository).save(R.string.wifi_band_key, WiFiBand.GHZ5.toggle().ordinal());
    }

    @Test
    public void testGetCountryCode() {
        // setup
        String defaultValue = LocaleUtils.getDefaultCountryCode();
        String expected = "WW";
        when(repository.getString(R.string.country_code_key, defaultValue)).thenReturn(expected);
        // execute
        String actual = fixture.getCountryCode();
        // validate
        assertEquals(expected, actual);
        verify(repository).getString(R.string.country_code_key, defaultValue);
    }

    @Test
    public void testGetLanguageLocale() {
        // setup
        String defaultValue = LocaleUtils.getDefaultLanguageTag();
        Locale expected = Locale.FRENCH;
        when(repository.getString(R.string.language_key, defaultValue)).thenReturn(LocaleUtils.toLanguageTag(expected));
        // execute
        Locale actual = fixture.getLanguageLocale();
        // validate
        assertEquals(expected, actual);
        verify(repository).getString(R.string.language_key, defaultValue);
    }

    @Test
    public void testGetSelectedMenu() {
        // setup
        when(repository.getStringAsInteger(R.string.selected_menu_key, NavigationMenu.ACCESS_POINTS.ordinal())).thenReturn(NavigationMenu.CHANNEL_GRAPH.ordinal());
        // execute
        NavigationMenu actual = fixture.getSelectedMenu();
        // validate
        assertEquals(NavigationMenu.CHANNEL_GRAPH, actual);
        verify(repository).getStringAsInteger(R.string.selected_menu_key, NavigationMenu.ACCESS_POINTS.ordinal());
    }

    @Test
    public void testSaveSelectedMenu() {
        // execute
        fixture.saveSelectedMenu(NavigationMenu.CHANNEL_GRAPH);
        // validate
        verify(repository).save(R.string.selected_menu_key, NavigationMenu.CHANNEL_GRAPH.ordinal());
    }

    @Test
    public void testSaveSelectedMenuWithNotAllowedMenu() {
        // execute
        fixture.saveSelectedMenu(NavigationMenu.ABOUT);
        // validate
        verify(repository, never()).save(anyInt(), anyInt());
    }

    @Test
    public void testIsWiFiOffOnExit() {
        // setup
        when(repository.getResourceBoolean(R.bool.wifi_off_on_exit_default)).thenReturn(true);
        when(repository.getBoolean(R.string.wifi_off_on_exit_key, true)).thenReturn(true);
        // execute
        boolean actual = fixture.isWiFiOffOnExit();
        // validate
        assertTrue(actual);
        verify(repository).getBoolean(R.string.wifi_off_on_exit_key, true);
        verify(repository).getResourceBoolean(R.bool.wifi_off_on_exit_default);
    }

    @Test
    public void testIsKeepScreenOn() {
        // setup
        when(repository.getResourceBoolean(R.bool.keep_screen_on_default)).thenReturn(true);
        when(repository.getBoolean(R.string.keep_screen_on_key, true)).thenReturn(true);
        // execute
        boolean actual = fixture.isKeepScreenOn();
        // validate
        assertTrue(actual);
        verify(repository).getBoolean(R.string.keep_screen_on_key, true);
        verify(repository).getResourceBoolean(R.bool.keep_screen_on_default);
    }
}

File: app/src/test/kotlin/com/vrem/wifianalyzer/settings/SettingsTest.kt
/*
 * WiFiAnalyzer
 * Copyright (C) 2015 - 2026 VREM Software Development <VREMSoftwareDevelopment@gmail.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>
 */
package com.vrem.wifianalyzer.settings

import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.os.Build
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.vrem.util.defaultCountryCode
import com.vrem.util.defaultLanguageTag
import com.vrem.util.ordinals
import com.vrem.util.toLanguageTag
import com.vrem.wifianalyzer.R
import com.vrem.wifianalyzer.navigation.NavigationMenu
import com.vrem.wifianalyzer.wifi.accesspoint.AccessPointViewType
import com.vrem.wifianalyzer.wifi.accesspoint.ConnectionViewType
import com.vrem.wifianalyzer.wifi.band.WiFiBand
import com.vrem.wifianalyzer.wifi.model.GroupBy
import com.vrem.wifianalyzer.wifi.model.Security
import com.vrem.wifianalyzer.wifi.model.SortBy
import com.vrem.wifianalyzer.wifi.model.Strength
import org.assertj.core.api.Assertions.assertThat
import org.junit.After
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.doNothing
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.mock
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.whenever
import org.robolectric.annotation.Config
import java.util.Locale

@RunWith(AndroidJUnit4::class)
@Config(sdk = [Build.VERSION_CODES.CINNAMON_BUN])
class SettingsTest {
    private val scanSpeedDefault = 5
    private val graphYMultiplier = -10
    private val graphYDefault = 2

    private val repository: Repository = mock()
    private val onSharedPreferenceChangeListener: OnSharedPreferenceChangeListener = mock()
    private val fixture = Settings(repository)

    @After
    fun tearDown() {
        verifyNoMoreInteractions(repository)
        verifyNoMoreInteractions(onSharedPreferenceChangeListener)
    }

    @Test
    fun initializeDefaultValues() {
        // Arrange
        doNothing().whenever(repository).initializeDefaultValues()
        // Act
        fixture.initializeDefaultValues()
        // Assert
        verify(repository).initializeDefaultValues()
    }

    @Test
    fun registerOnSharedPreferenceChangeListener() {
        // Arrange
        doNothing().whenever(repository).registerOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener)
        // Act
        fixture.registerOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener)
        // Assert
        verify(repository).registerOnSharedPreferenceChangeListener(onSharedPreferenceChangeListener)
    }

    @Test
    fun scanSpeed() {
        // Arrange
        val defaultValue = scanSpeedDefault - 2
        val speedValue = scanSpeedDefault - 1
        doReturn(defaultValue).whenever(repository).stringAsInteger(R.string.scan_speed_default, scanSpeedDefault)
        doReturn(speedValue).whenever(repository).stringAsInteger(R.string.scan_speed_key, defaultValue)
        // Act
        val actual = fixture.scanSpeed()
        // Assert
        assertThat(actual).isEqualTo(speedValue)
        verify(repository).stringAsInteger(R.string.scan_speed_default, scanSpeedDefault)
        verify(repository).stringAsInteger(R.string.scan_speed_key, defaultValue)
    }

    @Test
    fun graphMaximumY() {
        // Arrange
        val defaultValue = 1
        val value = 2
        val expected = value * graphYMultiplier
        doReturn(defaultValue).whenever(repository).stringAsInteger(R.string.graph_maximum_y_default, graphYDefault)
        doReturn(value).whenever(repository).stringAsInteger(R.string.graph_maximum_y_key, defaultValue)
        // Act
        val actual = fixture.graphMaximumY()
        // Assert
        assertThat(actual).isEqualTo(expected)
        verify(repository).stringAsInteger(R.string.graph_maximum_y_default, graphYDefault)
        verify(repository).stringAsInteger(R.string.graph_maximum_y_key, defaultValue)
    }

    @Test
    fun groupBy() {
        // Arrange
        doReturn(GroupBy.CHANNEL.ordinal)
            .whenever(repository)
            .stringAsInteger(R.string.group_by_key, GroupBy.NONE.ordinal)
        // Act
        val actual = fixture.groupBy()
        // Assert
        assertThat(actual).isEqualTo(GroupBy.CHANNEL)
        verify(repository).stringAsInteger(R.string.group_by_key, GroupBy.NONE.ordinal)
    }

    @Test
    fun sortBy() {
        // Arrange
        doReturn(SortBy.SSID.ordinal)
            .whenever(repository)
            .stringAsInteger(R.string.sort_by_key, SortBy.STRENGTH.ordinal)
        // Act
        val actual = fixture.sortBy()
        // Assert
        assertThat(actual).isEqualTo(SortBy.SSID)
        verify(repository).stringAsInteger(R.string.sort_by_key, SortBy.STRENGTH.ordinal)
    }

    @Test
    fun accessPointView() {
        // Arrange
        doReturn(AccessPointViewType.COMPACT.ordinal)
            .whenever(repository)
            .stringAsInteger(R.string.ap_view_key, AccessPointViewType.COMPLETE.ordinal)
        // Act
        val actual = fixture.accessPointView()
        // Assert
        assertThat(actual).isEqualTo(AccessPointViewType.COMPACT)
        verify(repository).stringAsInteger(R.string.ap_view_key, AccessPointViewType.COMPLETE.ordinal)
    }

    @Test
    fun connectionViewType() {
        // Arrange
        doReturn(ConnectionViewType.COMPLETE.ordinal)
            .whenever(repository)
            .stringAsInteger(R.string.connection_view_key, ConnectionViewType.COMPACT.ordinal)
        // Act
        val actual = fixture.connectionViewType()
        // Assert
        assertThat(actual).isEqualTo(ConnectionViewType.COMPLETE)
        verify(repository).stringAsInteger(R.string.connection_view_key, ConnectionViewType.COMPACT.ordinal)
    }

    @Test
    fun themeStyle() {
        ThemeStyle.entries.forEach {
            // Arrange
            doReturn(it.ordinal)
                .whenever(repository)
                .stringAsInteger(R.string.theme_key, ThemeStyle.DARK.ordinal)
            // Act
            val actual = fixture.themeStyle()
            // Assert
            assertThat(actual).describedAs("Theme: $it").isEqualTo(it)
        }
        verify(repository, times(ThemeStyle.entries.size))
            .stringAsInteger(R.string.theme_key, ThemeStyle.DARK.ordinal)
    }

    @Test
    fun themeStyleInvalid() {
        // Arrange
        doReturn(ThemeStyle.entries.size)
            .whenever(repository)
            .stringAsInteger(R.string.theme_key, ThemeStyle.DARK.ordinal)
        // Act
        val actual = fixture.themeStyle()
        // Assert
        assertThat(actual).isEqualTo(ThemeStyle.DARK)
        verify(repository).stringAsInteger(R.string.theme_key, ThemeStyle.DARK.ordinal)
    }

    @Test
    fun getWiFiBand() {
        // Arrange
        doReturn(WiFiBand.GHZ5.ordinal)
            .whenever(repository)
            .stringAsInteger(R.string.wifi_band_key, WiFiBand.GHZ2.ordinal)
        // Act
        val actual = fixture.wiFiBand()
        // Assert
        assertThat(actual).isEqualTo(WiFiBand.GHZ5)
        verify(repository).stringAsInteger(R.string.wifi_band_key, WiFiBand.GHZ2.ordinal)
    }

    @Test
    fun setWiFiBand() {
        // Arrange
        doNothing().whenever(repository).save(R.string.wifi_band_key, WiFiBand.GHZ5.ordinal)
        // Act
        fixture.wiFiBand(WiFiBand.GHZ5)
        // Assert
        verify(repository).save(R.string.wifi_band_key, WiFiBand.GHZ5.ordinal)
    }

    @Test
    fun settingsFindSSIDs() {
        // Arrange
        val expected: Set<String> = setOf("value1", "value2", "value3")
        doReturn(expected).whenever(repository).stringSet(R.string.filter_ssid_key, setOf())
        // Act
        val actual = fixture.findSSIDs()
        // Assert
        assertThat(actual).isEqualTo(expected)
        verify(repository).stringSet(R.string.filter_ssid_key, setOf())
    }

    @Test
    fun saveSSIDs() {
        // Arrange
        val values: Set<String> = setOf("value1", "value2", "value3")
        doNothing().whenever(repository).saveStringSet(R.string.filter_ssid_key, values)
        // Act
        fixture.saveSSIDs(values)
        // Assert
        verify(repository).saveStringSet(R.string.filter_ssid_key, values)
    }

    @Test
    fun settingsFindWiFiBands() {
        // Arrange
        val expected = WiFiBand.GHZ5
        val values = setOf("" + expected.ordinal)
        val defaultValues = ordinals(WiFiBand.entries)
        doReturn(values).whenever(repository).stringSet(R.string.filter_wifi_band_key, defaultValues)
        // Act
        val actual = fixture.findWiFiBands()
        // Assert
        assertThat(actual).hasSize(1)
        assertThat(actual).contains(expected)
        verify(repository).stringSet(R.string.filter_wifi_band_key, defaultValues)
    }

    @Test
    fun saveWiFiBands() {
        // Arrange
        val values = setOf(WiFiBand.GHZ5)
        val expected = setOf("" + WiFiBand.GHZ5.ordinal)
        doNothing().whenever(repository).saveStringSet(R.string.filter_wifi_band_key, expected)
        // Act
        fixture.saveWiFiBands(values)
        // Assert
        verify(repository).saveStringSet(R.string.filter_wifi_band_key, expected)
    }

    @Test
    fun settingsFindStrengths() {
        // Arrange
        val expected = Strength.THREE
        val values = setOf("" + expected.ordinal)
        val defaultValues = ordinals(Strength.entries)
        doReturn(values).whenever(repository).stringSet(R.string.filter_strength_key, defaultValues)
        // Act
        val actual = fixture.findStrengths()
        // Assert
        assertThat(actual).hasSize(1)
        assertThat(actual).contains(expected)
        verify(repository).stringSet(R.string.filter_strength_key, defaultValues)
    }

    @Test
    fun saveStrengths() {
        // Arrange
        val values = setOf(Strength.TWO)
        val expected = setOf("" + Strength.TWO.ordinal)
        doNothing().whenever(repository).saveStringSet(R.string.filter_strength_key, expected)
        // Act
        fixture.saveStrengths(values)
        // Assert
        verify(repository).saveStringSet(R.string.filter_strength_key, expected)
    }

    @Test
    fun settingsFindSecurities() {
        // Arrange
        val expected = Security.WPA
        val values = setOf("" + expected.ordinal)
        val defaultValues = ordinals(Security.entries)
        doReturn(values).whenever(repository).stringSet(R.string.filter_security_key, defaultValues)
        // Act
        val actual = fixture.findSecurities()
        // Assert
        assertThat(actual).hasSize(1)
        assertThat(actual).contains(expected)
        verify(repository).stringSet(R.string.filter_security_key, defaultValues)
    }

    @Test
    fun saveSecurities() {
        // Arrange
        val values = setOf(Security.WEP)
        val expected = setOf("" + Security.WEP.ordinal)
        doNothing().whenever(repository).saveStringSet(R.string.filter_security_key, expected)
        // Act
        fixture.saveSecurities(values)
        // Assert
        verify(repository).saveStringSet(R.string.filter_security_key, expected)
    }

    @Test
    fun countryCode() {
        // Arrange
        val defaultValue = defaultCountryCode()
        val expected = "WW"
        doReturn(expected).whenever(repository).string(R.string.country_code_key, defaultValue)
        // Act
        val actual = fixture.countryCode()
        // Assert
        assertThat(actual).isEqualTo(expected)
        verify(repository).string(R.string.country_code_key, defaultValue)
    }

    @Test
    fun languageLocale() {
        // Arrange
        val defaultValue = defaultLanguageTag()
        val expected = Locale.FRENCH
        doReturn(toLanguageTag(expected)).whenever(repository).string(R.string.language_key, defaultValue)
        // Act
        val actual = fixture.languageLocale()
        // Assert
        assertThat(actual).isEqualTo(expected)
        verify(repository).string(R.string.language_key, defaultValue)
    }

    @Test
    fun selectedMenu() {
        // Arrange
        doReturn(NavigationMenu.CHANNEL_GRAPH.ordinal)
            .whenever(repository)
            .stringAsInteger(R.string.selected_menu_key, NavigationMenu.ACCESS_POINTS.ordinal)
        // Act
        val actual = fixture.selectedMenu()
        // Assert
        assertThat(actual).isEqualTo(NavigationMenu.CHANNEL_GRAPH)
        verify(repository).stringAsInteger(R.string.selected_menu_key, NavigationMenu.ACCESS_POINTS.ordinal)
    }

    @Test
    fun saveSelectedMenu() {
        // Arrange
        doNothing().whenever(repository).save(R.string.selected_menu_key, NavigationMenu.CHANNEL_GRAPH.ordinal)
        // Act
        fixture.saveSelectedMenu(NavigationMenu.CHANNEL_GRAPH)
        // Assert
        verify(repository).save(R.string.selected_menu_key, NavigationMenu.CHANNEL_GRAPH.ordinal)
    }

    @Test
    fun saveSelectedMenuWithNotAllowedMenu() {
        // Act
        fixture.saveSelectedMenu(NavigationMenu.ABOUT)
    }

    @Test
    fun wiFiOffOnExit() {
        // Act
        val actual = fixture.wiFiOffOnExit()
        // Assert
        assertThat(actual).isFalse
    }

    @Test
    @Config(sdk = [Build.VERSION_CODES.P])
    fun wiFiOffOnExitLegacy() {
        // Arrange
        doReturn(true).whenever(repository).resourceBoolean(R.bool.wifi_off_on_exit_default)
        doReturn(true).whenever(repository).boolean(R.string.wifi_off_on_exit_key, true)
        // Act
        val actual = fixture.wiFiOffOnExit()
        // Assert
        assertThat(actual).isTrue
        verify(repository).boolean(R.string.wifi_off_on_exit_key, true)
        verify(repository).resourceBoolean(R.bool.wifi_off_on_exit_default)
    }

    @Test
    fun keepScreenOn() {
        // Arrange
        doReturn(true).whenever(repository).resourceBoolean(R.bool.keep_screen_on_default)
        doReturn(true).whenever(repository).boolean(R.string.keep_screen_on_key, true)
        // Act
        val actual = fixture.keepScreenOn()
        // Assert
        assertThat(actual).isTrue
        verify(repository).boolean(R.string.keep_screen_on_key, true)
        verify(repository).resourceBoolean(R.bool.keep_screen_on_default)
    }

    @Test
    fun cacheOff() {
        // Arrange
        doReturn(true).whenever(repository).resourceBoolean(R.bool.cache_off_default)
        doReturn(true).whenever(repository).boolean(R.string.cache_off_key, true)
        // Act
        val actual = fixture.cacheOff()
        // Assert
        assertThat(actual).isTrue
        verify(repository).boolean(R.string.cache_off_key, true)
        verify(repository).resourceBoolean(R.bool.cache_off_default)
    }
}
